### PR TITLE
Add graph topology visualization

### PR DESCRIPTION
**Pull request by johnwlambert**
## Deep Front End AstroNet
Input:
view_graph_estimator_input_topology_gt_locations

Output:
view_graph_estimator_output_topology_gt_locations

## Sift Front End AstroNet
Input:
view_graph_estimator_input_topology_gt_locations

Output:
view_graph_estimator_output_topology_gt_locations

## Skydio-32 Sift Front End
Input:
view_graph_estimator_input_topology_gt_locations

Output:
view_graph_estimator_output_topology_gt_locations

## Skydio-32 Deep Front End 
Input:
view_graph_estimator_input_topology_gt_locations

Output:
view_graph_estimator_output_topology_gt_locations

### CODE REVIEW

File: tests/utils/test_graph_utils.py [hunk 1]
```diff
@@ -229,6 +230,35 @@ def test_create_adjacency_list_empty(self) -> None:
         assert len(adj_list.keys()) == 0
         assert isinstance(adj_list, defaultdict)
 
+    def test_draw_graph_topology(self) -> None:
```
Comment (ayushbaid) [inline tests/utils/test_graph_utils.py:233]: This won't automatically test anything, right?

I suggest we mock the call to networkx, and verify the args passed to the mocked method..

**Review by ayushbaid — COMMENTED** [review]
Does this PR also plans to integrate this topology plots into the pipeline?

**Comment by johnwlambert** [comment]
> Does this PR also plans to integrate this topology plots into the pipeline?

I was thinking to add that in a separate PR, but could add it here too if you'd like.

File: gtsfm/utils/graph.py [hunk 2]
```diff
@@ -119,3 +126,60 @@ def extract_cyclic_triplets_from_edges(edges: List[Tuple[int, int]]) -> List[Tup
                 triplets.add(cycle_nodes)
 
     return list(triplets)
+
+
+def draw_graph_topology(
```
Comment (travisdriver) [inline gtsfm/utils/graph.py:131]: `draw_graph_topology` seems a little ambiguous; maybe `draw_view_graph` would be more descriptive?

**Comment by ayushbaid** [comment]
Also, having the nodes at their ground truth position might lead to weird visualizations depending on the actual post location? Like if some of them are too close to each other or stacked on top of each other?

I am fine with the current logic, just want to discuss briefly if there are other approaches.

**Comment by johnwlambert** [comment]
> Also, having the nodes at their ground truth position might lead to weird visualizations depending on the actual post location? Like if some of them are too close to each other or stacked on top of each other?
> 
> I am fine with the current logic, just want to discuss briefly if there are other approaches.

This is true. I've provided an option for Networkx to choose the node locations automatically, though (by passing in `cameras_gt=None`.

File: tests/utils/test_graph_utils.py (hunk 1 above)
Comment (johnwlambert) [reply tests/utils/test_graph_utils.py:233]: So this unit test is a "smokescreen" test IMHO. It makes sure that the code can run without crashing. It is a weak test, but I've inspected the plots and they look good.

File: gtsfm/utils/graph.py (hunk 2 above)
Comment (johnwlambert) [reply gtsfm/utils/graph.py]: Good point -- I renamed to `draw_view_graph_topology()`, let me know if that works.

**Comment by johnwlambert** [comment]
> This is really cool :) I also think it'd probably be cleaner to just integrate the plots into the pipeline in this PR as @ayushbaid mentioned.
> 
> It might also be more informative to color the edges depending on the magnitude of the pose error as opposed to a binary labeling with respect to a threshold.

I like the idea about coloring the edges based on the magnitude of the pose error. If we use a linear scale from [0,180], and we had 98 errors near 0 degree error, and 1 error near 20 degree error, and 1 error near 170 degrees error ,then the 20 degree error would be green in any colormap. So it would make sense maybe to do a uniform scale from [0,10], and then clamp. But unfortunately I don't have more time currently to play with that. I think we could add in a future PR.

**Comment by travisdriver** [comment]
> > This is really cool :) I also think it'd probably be cleaner to just integrate the plots into the pipeline in this PR as @ayushbaid mentioned.
> 
> > 
> 
> > It might also be more informative to color the edges depending on the magnitude of the pose error as opposed to a binary labeling with respect to a threshold.
> 
> 
> 
> I like the idea about coloring the edges based on the magnitude of the pose error. If we use a linear scale from [0,180], and we had 98 errors near 0 degree error, and 1 error near 20 degree error, and 1 error near 170 degrees error ,then the 20 degree error would be green in any colormap. So it would make sense maybe to do a uniform scale from [0,10], and then clamp. But unfortunately I don't have more time currently to play with that. I think we could add in a future PR.

I like the clamping idea, but yeah we can save it for a future PR. 

Everything else looks great to me.

**Review by travisdriver — APPROVED** [review]

**Comment by johnwlambert** [comment]
thanks for the review @travisdriver. it looks like the deep front end would do better on AstroNet if we can complete the loop (exhaustive matching), so maybe we should upload the cache of precomputed superpoint/superglue for RC3 to the CI? Have we done that yet?

**Pull request merged**